internal/setup/setup.go

```diff
@@ -81,79 +94,96 @@ func Setup(ctx context.Context, config DatabaseConfigProvider) (*serverenv.Serve
+		// Update the mutators to process secrets.
+		mutatorFuncs = append(mutatorFuncs, secrets.Resolver(sm, smConfig))
 
-	// Start building serverenv opts
-	opts := []serverenv.Option{
-		serverenv.WithSecretManager(sm),
-		serverenv.WithMetricsExporter(metrics.NewLogsBasedFromContext),
+		// Update serverEnv setup.
+		serverEnvOpts = append(serverEnvOpts, serverenv.WithSecretManager(sm))
 	}
 
-	// Configure key management for signing.
+	// Load the key manager.
+	var km signing.KeyManager
 	if provider, ok := config.(KeyManagerConfigProvider); ok {
 		kmConfig := provider.KeyManagerConfig()
-		km, err := signing.KeyManagerFor(ctx, kmConfig.KeyManagerType)
+		if err := envconfig.ProcessWith(ctx, kmConfig, l, mutatorFuncs...); err != nil {
+			return nil, fmt.Errorf("unable to process key manager env: %w", err)
+		}
+
+		var err error
+		km, err = signing.KeyManagerFor(ctx, kmConfig.KeyManagerType)
 		if err != nil {
-			return nil, nil, fmt.Errorf("unable to connect to key manager: %w", err)
+			return nil, fmt.Errorf("unable to connect to key manager: %w", err)
 		}
-		opts = append(opts, serverenv.WithKeyManager(km))
+
+		// Update serverEnv setup.
+		serverEnvOpts = append(serverEnvOpts, serverenv.WithKeyManager(km))
+	}
+
+	// Process first round of environment variables.
+	if err := envconfig.ProcessWith(ctx, config, l, mutatorFuncs...); err != nil {
+		return nil, fmt.Errorf("error loading environment variables: %w", err)
 	}
+	logger.Infof("Effective environment variables: %+v", config)
 
 	// Configure blob storage.
 	if provider, ok := config.(BlobstoreConfigProvider); ok {
 		bsConfig := provider.BlobstoreConfig()
 		blobStore, err := storage.BlobstoreFor(ctx, bsConfig.BlobstoreType)
 		if err != nil {
-			return nil, nil, fmt.Errorf("unable to connect to storage system: %v", err)
+			return nil, fmt.Errorf("unable to connect to storage system: %v", err)
 		}
 		blobStorage := serverenv.WithBlobStorage(blobStore)
-		opts = append(opts, blobStorage)
-	}
 
-	// Setup the database connection.
-	db, err := database.NewFromEnv(ctx, config.DatabaseConfig())
-	if err != nil {
-		return nil, nil, fmt.Errorf("unable to connect to database: %v", err)
-	}
-	{
-		// Log the database config, but omit the password field.
-		redactedDB := config.DatabaseConfig()
-		redactedDB.Password = "<hidden>"
-		logger.Infof("Effective DB config: %+v", redactedDB)
+		// Update serverEnv setup.
+		serverEnvOpts = append(serverEnvOpts, blobStorage)
 	}
-	opts = append(opts, serverenv.WithDatabase(db))
 
-	// AuthorizedApp must come after database setup due to the dependency.
-	if typ, ok := config.(AuthorizedAppConfigProvider); ok {
-		logger.Infof("Effective AuthorizedApp config: %+v", typ.AuthorizedAppConfig())
-		provider, err := authorizedapp.NewDatabaseProvider(ctx, db, typ.AuthorizedAppConfig(), authorizedapp.WithSecretManager(sm))
+	// Setup the database connection.
+	if provider, ok := config.(DatabaseConfigProvider); ok {
```

Comment:
one thing we have lost here, and I'm not sure if it is important or not, is that previously being a DatabaseConfigProvider provider was required (compile time interface check). Not, setup will be successful without one.

Reply:
Do you think that's a blocker? Happy to add it back, but it felt weird to accept a typed interface and then assert other types in the function body.

Reply:
let's go with this.